File: reference/app/recovery/cascade.py
"""Recovery Action Dry-Run Cascade — 反向影响范围计算。

输入:动作模板 + 目标资源
输出:受影响资源列表 + 严重度 + 估算 SLA

设计借鉴 `app.datasource.fault_injector._find_blast_targets` 的图遍历思路,
但有两个本质区别:

1. **不修改图状态** — fault_injector 注入故障会改 health/risk;cascade 只读不写
2. **多规则叠加** — 一个动作的 propagation 列表可能含 N 条规则,每条沿不同关系遍历;
   结果合并并去重

复用 `app.datasource.store` 作为图源(运行时孪生体),不直接查 Neo4j——
这样 dry-run 响应可以 < 100ms。
"""

from typing import Optional

from app.datasource.store import store
from app.datasource.models import DataNode
from app.recovery.action_defs import ACTION_DEFS, get_action


# 严重度优先级:high > medium > low > minimal(用于多规则命中同一节点时取最大值)
_IMPACT_RANK = {"minimal": 0, "low": 1, "medium": 2, "high": 3}
_REVERSE_IMPACT_RANK = {v: k for k, v in _IMPACT_RANK.items()}


def dry_run(action_id: str, target_resource_id: str,
            input_params: Optional[dict] = None) -> dict:
    """对一个 (动作 + 目标) 计算影响范围。

    返回结构:
    {
        "action_id": "scale_deployment",
        "target_resource_id": "deploy:...",
        "target_valid": True,
        "validation_error": None,
        "affected_resources": [
            {"resource_id": "...", "type": "Pod", "name": "...",
             "impact_severity": "minimal",
             "via_relations": ["CONTAINS"],
             "notes": ["新增/减少 Pod 副本"]},
            ...
        ],
        "estimated_duration_seconds": 90,
        "estimated_sla_impact": "< 0.1%",
        "warnings": [...],
        "rollback_action_id": "scale_deployment",
        "rollback_input_params": {"replicas_delta": -1},
    }

    若动作不存在或目标类型不匹配,target_valid=False,返回原因不抛异常
    (这样前端可以友好提示)。
    """
    action = get_action(action_id)
    if action is None:
        return _invalid_result(action_id, target_resource_id,
                               f"unknown action_id: {action_id}")

    target_node = store.get_node(target_resource_id)
    if target_node is None:
        return _invalid_result(action_id, target_resource_id,
                               f"target resource not found in DSS: {target_resource_id}",
                               action=action)

    if target_node.type != action["target_type"]:
        return _invalid_result(action_id, target_resource_id,
                               f"action targets {action['target_type']} but resource is {target_node.type}",
                               action=action)

    # 沿每条 propagation 规则遍历,合并结果
    affected: dict[str, dict] = {}    # resource_id -> impact dict
    for rule in action.get("propagation", []):
        for hit in _walk(target_node, rule):
            rid = hit["resource_id"]
            existing = affected.get(rid)
            if existing is None:
                affected[rid] = hit
            else:
                # 严重度取较大值
                if _IMPACT_RANK[hit["impact_severity"]] > _IMPACT_RANK[existing["impact_severity"]]:
                    existing["impact_severity"] = hit["impact_severity"]
                # 关系类型 / 注释合并
                for rel in hit["via_relations"]:
                    if rel not in existing["via_relations"]:
                        existing["via_relations"].append(rel)
                for note in hit["notes"]:
                    if note not in existing["notes"]:
                        existing["notes"].append(note)

    # 排除自身(目标节点)
    affected.pop(target_resource_id, None)

    # 按严重度倒序 + resource_id 排序,稳定输出
    affected_list = sorted(
        affected.values(),
        key=lambda x: (-_IMPACT_RANK[x["impact_severity"]], x["resource_id"]),
    )

    # 计算回滚参数(如果是 scale_deployment 的反向 delta)
    rollback_params = _compute_rollback_params(action, input_params or {})

    return {
        "action_id": action_id,
        "action_name": action["name"],
        "target_resource_id": target_resource_id,
        "target_resource_type": target_node.type,
        "target_resource_name": target_node.name,
        "target_valid": True,
        "validation_error": None,
        "affected_resources": affected_list,
        "affected_count": len(affected_list),
        "estimated_duration_seconds": action["estimated_duration_seconds"],
        "estimated_sla_impact": action["sla_impact_estimate"],
        "warnings": list(action.get("warnings", [])),
        "rollback_action_id": action.get("rollback_action_id"),
        "rollback_input_params": rollback_params,
        "risk_level": action["risk_level"],
        "requires_approval": action["requires_approval"],
    }


def _walk(start: DataNode, rule: dict) -> list[dict]:
    """从 start 节点出发沿 rule 描述的关系遍历,返回所有命中节点。

    BFS,避免深度遍历重复访问。
    """
    edge_type = rule["edge"]
    direction = rule["direction"]    # "forward" | "reverse"
    max_depth = rule.get("max_depth", 3)
    target_type = rule.get("target_type")
    impact = rule.get("impact", "low")
    note = rule.get("note", "")

    visited: set[str] = {start.id}
    frontier: list[str] = [start.id]
    hits: list[dict] = []

    for _ in range(max_depth):
        next_frontier: list[str] = []
        for node_id in frontier:
            for edge in store.get_all_edges():
                if edge.relationship_type != edge_type:
                    continue

                # 选下一跳:forward = source→target,reverse = target→source
                if direction == "forward" and edge.source_id == node_id:
                    next_id = edge.target_id
                elif direction == "reverse" and edge.target_id == node_id:
                    next_id = edge.source_id
                else:
                    continue

                if next_id in visited:
                    continue
                visited.add(next_id)
                next_frontier.append(next_id)

                # 类型筛选
                next_node = store.get_node(next_id)
                if next_node is None:
                    continue
                if target_type and next_node.type != target_type:
                    continue

                hits.append({
                    "resource_id": next_id,
                    "type": next_node.type,
                    "name": next_node.name,
                    "impact_severity": impact,
                    "via_relations": [edge_type],
                    "notes": [note] if note else [],
                })

        if not next_frontier:
            break
        frontier = next_frontier

    return hits


def _compute_rollback_params(action: dict, input_params: dict) -> dict | None:
    """对支持回滚的动作,计算回滚参数。

    Sprint 1 只处理 scale_deployment 的对称回滚:正负 delta 互换。
    其他动作的 rollback_action_id 为 None(不可逆)或 self(再 rollout 一次)。
    """
    rollback_id = action.get("rollback_action_id")
    if not rollback_id:
        return None

    if action["category"] == "scale" and "replicas_delta" in input_params:
        return {"replicas_delta": -input_params["replicas_delta"]}

    if rollback_id == action["name"] or rollback_id in ("rollback_deployment",):
        # 回滚动作自身没有简单参数(rollout undo 默认行为)
        return {}

    return None


def _invalid_result(action_id: str, target_resource_id: str,
                    error: str, action: dict | None = None) -> dict:
    """构造 target_valid=False 的响应。"""
    return {
        "action_id": action_id,
        "action_name": action["name"] if action else None,
        "target_resource_id": target_resource_id,
        "target_resource_type": None,
        "target_resource_name": None,
        "target_valid": False,
        "validation_error": error,
        "affected_resources": [],
        "affected_count": 0,
        "estimated_duration_seconds": action["estimated_duration_seconds"] if action else 0,
        "estimated_sla_impact": action["sla_impact_estimate"] if action else "n/a",
        "warnings": list(action.get("warnings", [])) if action else [],
        "rollback_action_id": action.get("rollback_action_id") if action else None,
        "rollback_input_params": None,
        "risk_level": action["risk_level"] if action else None,
        "requires_approval": action["requires_approval"] if action else None,
    }

File: engine/crates/engine-recovery/src/cascade.rs
//! Recovery Action Dry-Run Cascade - 影响范围计算(复刻 `reference/app/recovery/cascade.py`)。
//!
//! 输入:动作模板 + 目标资源 + 拓扑;输出:受影响资源列表 + 严重度 + 估算 SLA + 回滚参数。
//!
//! ## 与 reference 的差异
//!
//! - **I/O-free**:reference 读全局 DSS `store`;本模块把 [`engine_identity::Topology`]
//!   作入参传入(对齐 engine-identity 纯领域 + 薄持久化约定)。orchestration 层(Tauri/CLI)
//!   负责从 storage 取 materialized topology 喂进来。
//! - 算法逐字对齐:BFS `_walk`(forward/reverse + max_depth + target_type 筛选)、
//!   多规则命中取 max severity、关系/注释合并、排除自身、按 (-severity, resource_id) 排序。
//! - `rollback_input_params` 用 `serde_json::Value`(reference 用 dict)。

#![allow(missing_docs)]

use std::collections::{BTreeMap, HashMap, HashSet};

use engine_identity::{ResolvedNode, Topology};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::action_defs::{get_action, ActionDef, Direction, Impact, RiskLevel};

/// 一个受影响资源(merged 后)。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AffectedResource {
    /// 资源 ID。
    pub resource_id: String,
    /// 资源类型。JSON key `type`(对齐 reference)。
    #[serde(rename = "type")]
    pub type_: String,
    /// 资源名(label)。
    pub name: String,
    /// 影响严重度。
    pub impact_severity: Impact,
    /// 命中此节点的关系类型(去重合并)。
    pub via_relations: Vec<String>,
    /// 影响说明(去重合并)。
    pub notes: Vec<String>,
}

/// dry-run 结果。
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DryRunResult {
    /// 动作 ID。
    pub action_id: String,
    /// 动作名(动作未知时 None)。
    pub action_name: Option<String>,
    /// 目标资源 ID。
    pub target_resource_id: String,
    /// 目标资源类型(目标无效时 None)。
    pub target_resource_type: Option<String>,
    /// 目标资源名(目标无效时 None)。
    pub target_resource_name: Option<String>,
    /// 目标是否有效(动作存在 + 目标在拓扑 + 类型匹配)。
    pub target_valid: bool,
    /// 目标无效时的原因(target_valid=true 时 None)。
    pub validation_error: Option<String>,
    /// 受影响资源列表(按严重度倒序 + resource_id 升序)。
    pub affected_resources: Vec<AffectedResource>,
    /// 受影响资源数。
    pub affected_count: usize,
    /// 预估耗时(秒)。
    pub estimated_duration_seconds: u32,
    /// 预估 SLA 影响。
    pub estimated_sla_impact: String,
    /// 警告。
    pub warnings: Vec<String>,
    /// 回滚动作 ID(不可逆时 None)。
    pub rollback_action_id: Option<String>,
    /// 回滚参数(scale_deployment 反向 delta / rollback_deployment 空 {})。
    pub rollback_input_params: Option<Value>,
    /// 风险等级(动作未知时 None)。
    pub risk_level: Option<RiskLevel>,
    /// 是否需审批(动作未知时 None)。
    pub requires_approval: Option<bool>,
}

/// 对一个 (动作 + 目标) 计算影响范围。
///
/// 动作不存在 / 目标不在拓扑 / 类型不匹配 -> `target_valid=false` + `validation_error`
/// (不抛异常,前端可友好提示,对齐 reference)。
pub fn dry_run(
    action_id: &str,
    target_resource_id: &str,
    input_params: &Value,
    topology: &Topology,
) -> DryRunResult {
    let Some(action) = get_action(action_id) else {
        return invalid_result(action_id, target_resource_id, None, &format!("unknown action_id: {action_id}"));
    };

    // 节点索引:resource_id -> &ResolvedNode
    let node_index: HashMap<&str, &ResolvedNode> = topology
        .nodes
        .iter()
        .map(|n| (n.resource_id.as_str(), n))
        .collect();

    let Some(target_node) = node_index.get(target_resource_id) else {
        return invalid_result(
            action_id,
            target_resource_id,
            Some(action),
            &format!("target resource not found in topology: {target_resource_id}"),
        );
    };

    if target_node.resource_type != action.target_type {
        return invalid_result(
            action_id,
            target_resource_id,
            Some(action),
            &format!(
                "action targets {} but resource is {}",
                action.target_type, target_node.resource_type
            ),
        );
    }

    // 沿每条 propagation 规则 BFS,合并结果
    let mut affected: BTreeMap<String, AffectedResource> = BTreeMap::new();
    for rule in action.propagation {
        for hit in walk(target_resource_id, rule, topology, &node_index) {
            match affected.get_mut(&hit.resource_id) {
                Some(existing) => {
                    // 严重度取较大值
                    if hit.impact.rank() > existing.impact_severity.rank() {
                        existing.impact_severity = hit.impact;
                    }
                    // 关系 / 注释合并去重
                    if !existing.via_relations.contains(&hit.via_relation) {
                        existing.via_relations.push(hit.via_relation);
                    }
                    if !existing.notes.contains(&hit.note) {
                        existing.notes.push(hit.note);
                    }
                }
                None => {
                    affected.insert(
                        hit.resource_id.clone(),
                        AffectedResource {
                            resource_id: hit.resource_id,
                            type_: hit.type_,
                            name: hit.name,
                            impact_severity: hit.impact,
                            via_relations: vec![hit.via_relation],
                            notes: vec![hit.note],
                        },
                    );
                }
            }
        }
    }

    // 排除自身
    affected.remove(target_resource_id);

    // 按 (-severity, resource_id) 排序,稳定输出
    let mut affected_list: Vec<AffectedResource> = affected.into_values().collect();
    affected_list.sort_by(|a, b| {
        b.impact_severity
            .rank()
            .cmp(&a.impact_severity.rank())
            .then_with(|| a.resource_id.cmp(&b.resource_id))
    });

    let rollback_input_params = compute_rollback_params(action, input_params);

    DryRunResult {
        action_id: action_id.to_string(),
        action_name: Some(action.name.to_string()),
        target_resource_id: target_resource_id.to_string(),
        target_resource_type: Some(target_node.resource_type.clone()),
        target_resource_name: Some(target_node.label.clone()),
        target_valid: true,
        validation_error: None,
        affected_count: affected_list.len(),
        affected_resources: affected_list,
        estimated_duration_seconds: action.estimated_duration_seconds,
        estimated_sla_impact: action.sla_impact_estimate.to_string(),
        warnings: action.warnings.iter().map(|w| w.to_string()).collect(),
        rollback_action_id: action.rollback_action_id.map(|s| s.to_string()),
        rollback_input_params,
        risk_level: Some(action.risk_level),
        requires_approval: Some(action.requires_approval),
    }
}

/// BFS 一条 propagation 规则,返回所有命中节点(未合并)。
fn walk<'t>(
    start_id: &str,
    rule: &crate::action_defs::PropagationRule,
    topology: &'t Topology,
    node_index: &HashMap<&str, &'t ResolvedNode>,
) -> Vec<Hit> {
    let mut visited: HashSet<&str> = HashSet::new();
    visited.insert(start_id);
    let mut frontier: Vec<&str> = vec![start_id];
    let mut hits: Vec<Hit> = Vec::new();

    for _ in 0..rule.max_depth {
        let mut next_frontier: Vec<&str> = Vec::new();
        for node_id in &frontier {
            for edge in &topology.edges {
                if edge.edge_type != rule.edge {
                    continue;
                }
                // forward = source->target;reverse = target->source
                let next_id = match rule.direction {
                    Direction::Forward if edge.source == *node_id => edge.target.as_str(),
                    Direction::Reverse if edge.target == *node_id => edge.source.as_str(),
                    _ => continue,
                };
                if visited.contains(next_id) {
                    continue;
                }
                visited.insert(next_id);
                next_frontier.push(next_id);

                let Some(node) = node_index.get(next_id) else {
                    continue;
                };
                if let Some(tt) = rule.target_type {
                    if node.resource_type != tt {
                        continue;
                    }
                }
                hits.push(Hit {
                    resource_id: next_id.to_string(),
                    type_: node.resource_type.clone(),
                    name: node.label.clone(),
                    impact: rule.impact,
                    via_relation: rule.edge.to_string(),
                    note: rule.note.to_string(),
                });
            }
        }
        if next_frontier.is_empty() {
            break;
        }
        frontier = next_frontier;
    }
    hits
}

/// walk 的单条命中(合并前)。
struct Hit {
    resource_id: String,
    type_: String,
    name: String,
    impact: Impact,
    via_relation: String,
    note: String,
}

/// 计算回滚参数(对齐 reference `_compute_rollback_params`)。
///
/// - scale_deployment(replicas_delta=N) -> `{"replicas_delta": -N}`
/// - rollback_deployment -> `{}`(rollout undo 默认行为)
/// - 其它 -> None(不可逆或 self-rollback 无简单参数)
fn compute_rollback_params(action: &ActionDef, input_params: &Value) -> Option<Value> {
    let rollback_id = action.rollback_action_id?;
    if action.category == "scale" {
        if let Some(delta) = input_params.get("replicas_delta").and_then(Value::as_i64) {
            return Some(serde_json::json!({ "replicas_delta": -delta }));
        }
    }
    if rollback_id == "rollback_deployment" {
        return Some(serde_json::json!({}));
    }
    None
}

/// 构造 `target_valid=false` 的结果。
fn invalid_result(
    action_id: &str,
    target_resource_id: &str,
    action: Option<&ActionDef>,
    error: &str,
) -> DryRunResult {
    DryRunResult {
        action_id: action_id.to_string(),
        action_name: action.map(|a| a.name.to_string()),
        target_resource_id: target_resource_id.to_string(),
        target_resource_type: None,
        target_resource_name: None,
        target_valid: false,
        validation_error: Some(error.to_string()),
        affected_resources: vec![],
        affected_count: 0,
        estimated_duration_seconds: action.map(|a| a.estimated_duration_seconds).unwrap_or(0),
        estimated_sla_impact: action
            .map(|a| a.sla_impact_estimate.to_string())
            .unwrap_or_else(|| "n/a".to_string()),
        warnings: action
            .map(|a| a.warnings.iter().map(|w| w.to_string()).collect())
            .unwrap_or_default(),
        rollback_action_id: action.and_then(|a| a.rollback_action_id.map(|s| s.to_string())),
        rollback_input_params: None,
        risk_level: action.map(|a| a.risk_level),
        requires_approval: action.map(|a| a.requires_approval),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use engine_identity::{ResolvedEdge, ResolvedNode, Topology};
    use serde_json::json;

    /// 构造 reference test_recovery.py 的 fixture 拓扑(9 节点 11 边)。
    ///
    /// ```text
    /// app:order
    ///   └ comp:order-api      (BELONGS_TO 反向 = comp 在 app 下)
    ///       └ deploy:order-api  (DEPLOYED_AS)
    ///           └ pod:order-api-1  (CONTAINS)
    ///           └ pod:order-api-2  (CONTAINS)
    ///               └ node:worker-1  (SCHEDULED_ON)
    /// svc:order-api  (ROUTES_TO pod:order-api-1, pod:order-api-2)
    /// secret:order-tls  (USES -> pod:order-api-1)
    /// mysql:order-db  (USES -> pod:order-api-1)
    /// ```
    fn fixture_topology() -> Topology {
        let nodes = vec![
            node("app:order", "Application", "订单应用"),
            node("comp:order-api", "ApplicationComponent", "订单API组件"),
            node("deploy:order-api", "Deployment", "order-api"),
            node("pod:order-api-1", "Pod", "order-api-1"),
            node("pod:order-api-2", "Pod", "order-api-2"),
            node("node:worker-1", "KubernetesNode", "worker-1"),
            node("svc:order-api", "Service", "order-api-svc"),
            node("secret:order-tls", "Secret", "order-tls"),
            node("mysql:order-db", "MySQL", "order-db"),
        ];
        let edges = vec![
            edge("e1", "app:order", "CONTAINS", "comp:order-api"),
            edge("e2", "comp:order-api", "DEPLOYED_AS", "deploy:order-api"),
            edge("e3", "deploy:order-api", "CONTAINS", "pod:order-api-1"),
            edge("e4", "deploy:order-api", "CONTAINS", "pod:order-api-2"),
            edge("e5", "pod:order-api-2", "SCHEDULED_ON", "node:worker-1"),
            edge("e6", "svc:order-api", "ROUTES_TO", "pod:order-api-1"),
            edge("e7", "svc:order-api", "ROUTES_TO", "pod:order-api-2"),
            edge("e8", "pod:order-api-1", "USES", "secret:order-tls"),
            edge("e9", "pod:order-api-1", "USES", "mysql:order-db"),
            edge("e10", "deploy:order-api", "BELONGS_TO", "comp:order-api"),
            edge("e11", "comp:order-api", "BELONGS_TO", "app:order"),
        ];
        Topology { nodes, edges }
    }

    fn node(rid: &str, rtype: &str, label: &str) -> ResolvedNode {
        ResolvedNode {
            resource_id: rid.into(),
            resource_type: rtype.into(),
            label: label.into(),
            attributes_json: "{}".into(),
        }
    }

    fn edge(id: &str, src: &str, etype: &str, tgt: &str) -> ResolvedEdge {
        ResolvedEdge {
            id: id.into(),
            source: src.into(),
            target: tgt.into(),
            edge_type: etype.into(),
        }
    }

    fn affected_ids(r: &DryRunResult) -> HashSet<String> {
        r.affected_resources
            .iter()
            .map(|a| a.resource_id.clone())
            .collect()
    }

    #[test]
    fn unknown_action_invalid() {
        let r = dry_run("nonexistent_action", "pod:x", &json!({}), &fixture_topology());
        assert!(!r.target_valid);
        assert!(r.validation_error.as_deref().unwrap().contains("unknown action_id"));
    }

    #[test]
    fn target_not_in_topology_invalid() {
        let r = dry_run("restart_pod", "pod:does-not-exist", &json!({}), &fixture_topology());
        assert!(!r.target_valid);
        assert!(r.validation_error.as_deref().unwrap().contains("not found"));
    }

    #[test]
    fn target_type_mismatch_invalid() {
        // restart_pod targets Pod, but deploy:order-api is Deployment
        let r = dry_run("restart_pod", "deploy:order-api", &json!({}), &fixture_topology());
        assert!(!r.target_valid);
        let err = r.validation_error.as_deref().unwrap();
        assert!(err.contains("Pod"), "err should mention action target_type Pod: {err}");
        assert!(err.contains("Deployment"), "err should mention resource type Deployment: {err}");
    }

    #[test]
    fn scale_deployment_propagation() {
        let r = dry_run(
            "scale_deployment",
            "deploy:order-api",
            &json!({ "replicas_delta": 2 }),
            &fixture_topology(),
        );
        assert!(r.target_valid);
        let ids = affected_ids(&r);
        // forward CONTAINS -> 2 个 Pod
        assert!(ids.contains("pod:order-api-1"));
        assert!(ids.contains("pod:order-api-2"));
        // BELONGS_TO 链向上 -> component / application
        assert!(ids.contains("comp:order-api") || ids.contains("app:order"));
        // 自身排除
        assert!(!ids.contains("deploy:order-api"));
        // 回滚参数:反向 delta
        assert_eq!(r.rollback_input_params, Some(json!({ "replicas_delta": -2 })));
    }

    #[test]
    fn restart_pod_propagation() {
        // reverse ROUTES_TO -> Service
        let r = dry_run("restart_pod", "pod:order-api-1", &json!({}), &fixture_topology());
        assert!(r.target_valid);
        let ids = affected_ids(&r);
        assert!(ids.contains("svc:order-api"), "reverse ROUTES_TO should hit Service");
    }

    #[test]
    fn drain_node_propagation() {
        // reverse SCHEDULED_ON -> Pod 调度在此节点
        let r = dry_run("drain_node", "node:worker-1", &json!({}), &fixture_topology());
        assert!(r.target_valid);
        let ids = affected_ids(&r);
        assert!(ids.contains("pod:order-api-2"), "pod:order-api-2 scheduled on worker-1");
    }

    #[test]
    fn refresh_secret_propagation() {
        // reverse USES -> 引用此 Secret 的 Pod
        let r = dry_run("refresh_secret", "secret:order-tls", &json!({}), &fixture_topology());
        assert!(r.target_valid);
        let ids = affected_ids(&r);
        assert!(ids.contains("pod:order-api-1"), "reverse USES should hit pod using secret");
    }

    #[test]
    fn severity_aggregation_takes_max() {
        // rollback_deployment 的 Pod 是 medium(forward CONTAINS)
        let r = dry_run("rollback_deployment", "deploy:order-api", &json!({}), &fixture_topology());
        assert!(r.target_valid);
        let pods: Vec<&AffectedResource> = r
            .affected_resources
            .iter()
            .filter(|a| a.type_ == "Pod")
            .collect();
        assert!(!pods.is_empty());
        for p in &pods {
            assert!(
                p.impact_severity == Impact::Medium || p.impact_severity == Impact::High,
                "Pod impact should be >= medium, got {:?}",
                p.impact_severity
            );
        }
    }

    #[test]
    fn rollback_deployment_rollback_params_empty_object() {
        let r = dry_run("rollback_deployment", "deploy:order-api", &json!({}), &fixture_topology());
        assert!(r.target_valid);
        assert_eq!(r.rollback_input_params, Some(json!({})));
    }

    #[test]
    fn non_reversible_action_has_no_rollback_params() {
        // restart_pod rollback_action_id=None -> None
        let r = dry_run("restart_pod", "pod:order-api-1", &json!({}), &fixture_topology());
        assert!(r.target_valid);
        assert_eq!(r.rollback_input_params, None);
    }

    #[test]
    fn affected_resources_sorted_by_severity_then_id() {
        // refresh_secret 命中 Pod(medium,reverse USES)+ Deployment(low,reverse USES)
        // + BELONGS_TO 链(low)。验排序:severity 倒序,同 severity 按 resource_id 升序。
        let r = dry_run("refresh_secret", "secret:order-tls", &json!({}), &fixture_topology());
        assert!(r.target_valid);
        let sev = r
            .affected_resources
            .iter()
            .map(|a| a.impact_severity.rank())
            .collect::<Vec<_>>();
        let mut sorted = sev.clone();
        sorted.sort_by(|a, b| b.cmp(a));
        assert_eq!(sev, sorted, "affected should be sorted by severity desc: {:?}", sev);
    }
}
